waku/v2/protocol/waku_filter_v2/protocol.nim

```diff
@@ -0,0 +1,159 @@
+    return err(FilterSubscribeError.badRequest("pubsubTopic and contentTopics must be specified"))
+
+  let filterCriteria = toHashSet(contentTopics.mapIt((pubsubTopic.get(), it)))
+
+  trace "unsubscribing peer from filter criteria", peerId=peerId, filterCriteria=filterCriteria
+
+  if peerId notin wf.subscriptions:
+    debug "unsubscibing peer has no subscriptions", peerId=peerId
+    return err(FilterSubscribeError.notFound())
+
+  var peerSubscription = wf.subscriptions.mgetOrPut(peerId, initHashSet[FilterCriterion]())
+  # TODO: consider error response if filter criteria does not exist
+  peerSubscription.excl(filterCriteria)
+
+  if peerSubscription.len() == 0:
+    debug "peer has no more subscriptions, removing subscription", peerId=peerId
+    wf.subscriptions.del(peerId)
+  else:
+    wf.subscriptions[peerId] = peerSubscription
+
+  ok()
+
+proc unsubscribeAll(wf: WakuFilter, peerId: PeerID): FilterSubscribeResult =
+  if peerId notin wf.subscriptions:
+    debug "unsubscibing peer has no subscriptions", peerId=peerId
+    return err(FilterSubscribeError.notFound())
+
+  debug "removing peer subscription", peerId=peerId
+  wf.subscriptions.del(peerId)
+
+  ok()
+
+proc handleSubscribeRequest*(wf: WakuFilter, peerId: PeerId, request: FilterSubscribeRequest): FilterSubscribeResponse =
+  info "received filter subscribe request", peerId=peerId, request=request
+  waku_filter_requests.inc(labelValues = [$request.filterSubscribeType])
+
+  var subscribeResult: FilterSubscribeResult
+
+  case request.filterSubscribeType
+  of FilterSubscribeType.SUBSCRIBER_PING:
+    subscribeResult = wf.pingSubscriber(peerId)
+  of FilterSubscribeType.SUBSCRIBE:
+    subscribeResult = wf.subscribe(peerId, request.pubsubTopic, request.contentTopics)
+  of FilterSubscribeType.UNSUBSCRIBE:
+    subscribeResult = wf.unsubscribe(peerId, request.pubsubTopic, request.contentTopics)
+  of FilterSubscribeType.UNSUBSCRIBE_ALL:
+    subscribeResult = wf.unsubscribeAll(peerId)
+
+  if subscribeResult.isErr():
+    return FilterSubscribeResponse(
+      requestId: request.requestId,
+      statusCode: subscribeResult.error.kind.uint32,
+      statusDesc: some($subscribeResult.error)
+    )
+  else:
+    return FilterSubscribeResponse.ok(request.requestId)
+
+proc handleMessage*(wf: WakuFilter, message: WakuMessage) =
+  raiseAssert "Unimplemented"
+
+proc initProtocolHandler(wf: WakuFilter) =
+
+  proc handler(conn: Connection, proto: string) {.async.} =
+    let buf = await conn.readLp(MaxSubscribeSize)
+
+    let decodeRes = FilterSubscribeRequest.decode(buf)
+    if decodeRes.isErr():
+      error "Failed to decode filter subscribe request", peerId=conn.peerId
+      waku_filter_errors.inc(labelValues = [decodeRpcFailure])
+      return
+
+    let request = decodeRes.value #TODO: toAPI() split here
+
+    let response = wf.handleSubscribeRequest(conn.peerId, request)
+
+    await conn.writeLp(response.encode().buffer) #TODO: toRPC() separation here
```

Comment:
not very likely to happen in reality, but if the connection is closed by the counterparty right before, this will throw an exception. I just verified empirically when modifying px.

https://github.com/waku-org/nwaku/blob/master/tests/v2/test_waku_peer_exchange.nim#L247